### ISSUE TITLE
Support checkpoints in gradients

### ISSUE THREAD

**Issue by josevalim** [elixir-nx/nx#765]
Reference JAX implementation: https://github.com/google/jax/pull/1749

**Comment by josevalim** [comment]
https://dev-discuss.pytorch.org/t/min-cut-optimal-recomputation-i-e-activation-checkpointing-with-aotautograd/467

**Comment by blasphemetheus** [comment]
I'd like an axon feature blocked by this, so will start to look into this one, will try and post progress here

**Comment by blasphemetheus** [comment]
right, so I did a first try at this over at https://github.com/blasphemetheus/nx/tree/feat/gradient-checkpointing 

things to note:
- Nx.Defn.checkpoint(input, fun) — explicit inputs API, pipe-friendly
- :checkpoint expression node following the :optional pattern
- Evaluator backend support (pass-through with correct parameter resolution)
- Gradient via parents_args re-tracing (same as :optional)
- 52 tests
- design doc in nx/CHECKPOINT_DESIGN.md

not done: EXLA backend (needs optimization barrier [thing says don't optimize across this boundary] readded for Common Subexpression Elimination prevention - we want to recompute the backward pass from the forward pass not reuse the computed value from the forward pass), any sort of axon integration (#372), rematerialization policies (checkpoint recomputes everything inside the body rn, policy lets you decide, the future optimization that was mentioned)

design recipe made this a lot faster holy cow (did this while waiting for someone at an ER overnight :) )

**Comment by blasphemetheus** [comment]
gotcha -> added a test for the recomputation behavior and pushed a commit making that change on the fork branch

**Comment by blasphemetheus** [comment]
gotcha!

**Comment by blasphemetheus** [comment]
this is one of two open issues, seems like a better time to revisit this

**Comment by polvalente** [comment]
I agree. We'll probably leave this one for after the upcoming 1.0 release as it should be purely additive, but I think this is a missing feature for Axon 1.0. If you have thoughts on a more concrete implementation, we can discuss them!
